pkg/identity/cache.go

```diff
@@ -0,0 +1,128 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identity
+
+import (
+	"reflect"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/kvstore/allocator"
+	"github.com/cilium/cilium/pkg/labels"
+)
+
+var (
+	reservedIdentityCache = map[NumericIdentity]*Identity{}
+)
+
+// IdentityCache is a cache of identity to labels mapping
+type IdentityCache map[NumericIdentity]labels.LabelArray
+
+// GetIdentityCache returns a cache of all known identities
+func GetIdentityCache() IdentityCache {
+	cache := IdentityCache{}
+
+	identityAllocator.ForeachCache(func(id allocator.ID, val allocator.AllocatorKey) {
+		gi := val.(globalIdentity)
+		cache[NumericIdentity(id)] = gi.LabelArray()
+	})
+
+	return cache
+}
+
+// GetIdentities returns all known identities
+func GetIdentities() []*models.Identity {
+	identities := []*models.Identity{}
+
+	identityAllocator.ForeachCache(func(id allocator.ID, val allocator.AllocatorKey) {
+		if gi, ok := val.(globalIdentity); ok {
+			identity := NewIdentity(NumericIdentity(id), gi.Labels)
+			identities = append(identities, identity.GetModel())
+		}
+
+	})
+
+	return identities
+}
+
+func identityWatcher(owner IdentityAllocatorOwner) {
+	for {
+		event := <-identityAllocator.Events
+
+		switch event.Typ {
+		case kvstore.EventTypeCreate, kvstore.EventTypeDelete:
+			owner.TriggerPolicyUpdates(true)
+
+		case kvstore.EventTypeModify:
+			// Ignore modify events
+		}
+	}
+}
+
+// LookupIdentity looks up the identity by its labels but does not create it.
+// This function will first search through the local cache and fall back to
+// querying the kvstore.
+func LookupIdentity(lbls labels.Labels) *Identity {
+	for _, identity := range reservedIdentityCache {
+		if reflect.DeepEqual(identity.Labels, lbls) {
+			return identity
+		}
+	}
+
+	if identityAllocator == nil {
+		return nil
+	}
+
+	id, err := identityAllocator.Get(globalIdentity{lbls})
+	if err != nil {
+		return nil
+	}
+
+	if id == allocator.NoID {
+		return nil
+	}
+
+	return NewIdentity(NumericIdentity(id), lbls)
+}
+
+// LookupIdentityByID returns the identity by ID. This function will first
+// search through the local cache and fall back to querying the kvstore.
+func LookupIdentityByID(id NumericIdentity) *Identity {
+	if identity, ok := reservedIdentityCache[id]; ok {
+		return identity
+	}
+
+	if identityAllocator == nil {
+		return nil
+	}
+
+	allocatorKey, err := identityAllocator.GetByID(allocator.ID(id))
+	if err != nil {
+		return nil
+	}
+
+	if gi, ok := allocatorKey.(globalIdentity); ok {
+		return NewIdentity(id, gi.Labels)
```

Comment:
Why don't we store the updated identity from the KVStore into the cache?

Reply:
The lookup is using the cache in the allocator but falls back to kvstore if it can't be found in the cache. I'm just moving the code here, I'm not adding or modifying code.